### ISSUE TITLE
websocket connection error: cafile, capath and cadata cannot be all omitted

### ISSUE THREAD

**Issue by terlar** [rawdigits/wee-slack#122]
I got the following error when I try to load the slack plugin:

```
python: stdout/stderr: websocket connection error: cafile, capath and cadata cannot be all omitted
python: stdout/stderr: websocket connection error: cafile, capath and cadata cannot be all omitted
```

After this it just infinitely tries to reconnect to slack without success.

I think this might have been due to some upgrade of libraries but not sure. Any pointers in the right direction would be appreciated or of this is actually a bug in wee-slack.

**Comment by kenyonj** [comment]
I am seeing this issue as well... I am thinking that it was some python library that was updates, but I haven't figured out which one yet.

**Comment by kenyonj** [comment]
@terlar were you able to figure out the root cause of this issue? I have resorted to the web view of slack in the meantime, but I am very anxious to get back.

Can I ask what OS you are running?

**Comment by terlar** [comment]
@kenyonj No I have not yet looked into this, also resorting to the web UI.

I could potentially bring up a package install log to narrow suspects down. I am running Archlinux. What about you?

**Comment by pigmonkey** [comment]
I attempted to start using wee-slack yesterday and encountered this same error. I'm also on Arch and installed the [python2-websocket-client](https://www.archlinux.org/packages/community/any/python2-websocket-client/) prior to trying the wee-slack script.

Since 0.34 is a [recent release](https://projects.archlinux.org/svntogit/community.git/log/trunk?h=packages/python-websocket-client) of websocket-client I thought the problem may be with that, but I received the same error after downgrading to both 0.33 and 0.32.

Finally I just created a new virtualenv, installed websocket-client, and ran weechat within that. That works fine, so I have no idea what system library is causing the error.

    $ pip freeze
    six==1.10.0
    websocket-client==0.34.0
    wheel==0.24.0

I did create the virtualenv with python2, rather than Arch's default of python3. Maybe its a 2 vs 3 thing?

**Comment by kenyonj** [comment]
Looks like we are all on Arch... I will try your fix @pigmonkey, thanks for the details.

**Comment by bchretien** [comment]
In your virtualenv, could you try:
```sh
$ wsdump2.py wss://ms332.slack-msgs.com
```

On Arch, this currently returns the `cafile, capath and cadata cannot be all omitted` error.

**Comment by pigmonkey** [comment]
In my virtualenv:

```
$ wsdump.py wss://ms332.slack-msgs.com
Handshake status 200
```

**Comment by terlar** [comment]
I get the same error as you (not using virtualenv). Not really familiar with the python stuff so haven't bothered with setting up a virtual env.

**Comment by jferris** [comment]
I'm not familiar with Python either, but this worked for me on Arch:

```
sudo pacman -S install python2 python2-virtualenv
virtualenv2 weechatpython2
cd weechatpython2
source bin/activate
pip install pync
pip install websocket-client
weechat
```

**Comment by etu** [comment]
So, anyone planning a real solution instead of doing workarounds to install an old version? Also on arch, got this issue.

The real solution would be to actually fix the problem. Other distros will probably receive newer packages as well at some point.

**Comment by petr-fischer** [comment]
I also don't want "virtualenv mess" on my Arch. Unfortunately, I can't fix this issue myself :(

**Comment by etu** [comment]
Yeah, I tried to build an old version of the websocket-client package and install it with the plan to configure pacman to not upgrade that package (which is doable in `pacman.conf`).

But unfortunately, this didn't work for me. I still got the same error after `/python reload`.

**Comment by daurnimator** [comment]
Same issue here (also running Arch)

**Comment by daurnimator** [comment]
This patch fixes it:
```
diff --git a/wee_slack.py b/wee_slack.py
index c6d9b40..256fead 100644
--- a/wee_slack.py
+++ b/wee_slack.py
@@ -14,6 +14,10 @@ import sys
 import traceback
 import collections
 from websocket import create_connection,WebSocketConnectionClosedException
+import ssl
+
+ssl_defaults = ssl.get_default_verify_paths()
+sslopt_ca_certs = {'ca_certs': ssl_defaults.cafile}
 
 # hack to make tests possible.. better way?
 try:
@@ -261,7 +265,7 @@ class SlackServer(object):
     def create_slack_websocket(self, data):
         web_socket_url = data['url']
         try:
-            self.ws = create_connection(web_socket_url)
+            self.ws = create_connection(web_socket_url, sslopt=sslopt_ca_certs)
             self.ws_hook = w.hook_fd(self.ws.sock._sock.fileno(), 1, 0, 0, "slack_websocket_cb", self.identifier)
             self.ws.sock.setblocking(0)
             return True
01-31 04:57:15 daurnimator@daurn-vultr ~/wee-slack git:{master...origin/master:8a9363b} $ git diff | xclip
-bash: xclip: command not found
01-31 04:57:17 daurnimator@daurn-vultr ~/wee-slack git:{master...origin/master:8a9363b} exit=127 $ git diff | xclip
-bash: xclip: command not found
01-31 04:57:25 daurnimator@daurn-vultr ~/wee-slack git:{master...origin/master:8a9363b} exit=127 $ git diff 
diff --git a/wee_slack.py b/wee_slack.py
index c6d9b40..256fead 100644
--- a/wee_slack.py
+++ b/wee_slack.py
@@ -14,6 +14,10 @@ import sys
 import traceback
 import collections
 from websocket import create_connection,WebSocketConnectionClosedException
+import ssl
+
+ssl_defaults = ssl.get_default_verify_paths()
+sslopt_ca_certs = {'ca_certs': ssl_defaults.cafile}
 
 # hack to make tests possible.. better way?
 try:
@@ -261,7 +265,7 @@ class SlackServer(object):
     def create_slack_websocket(self, data):
         web_socket_url = data['url']
         try:
-            self.ws = create_connection(web_socket_url)
+            self.ws = create_connection(web_socket_url, sslopt=sslopt_ca_certs)
             self.ws_hook = w.hook_fd(self.ws.sock._sock.fileno(), 1, 0, 0, "slack_websocket_cb", self.identifier)
             self.ws.sock.setblocking(0)
             return True
```

**Comment by rawdigits** [comment]
Hi @daurnimator - Thank you for fixing this for lots of folks. my irc VM must be rather outdated, as ssl.get_default_verify_paths doesn't exist for me. I'm happy to implement your changes once I know how to make them backward compatible. Would a `try:` be best, or a version check? Thanks for any advice you can share!

**Comment by trygveaa** [comment]
@rawdigits: I posted a patch for making it backwards compatible in the PR.

**Comment by jdearmas** [comment]
> I'm not familiar with Python either, but this worked for me on Arch:
> 
> ```
> sudo pacman -S install python2 python2-virtualenv
> virtualenv2 weechatpython2
> cd weechatpython2
> source bin/activate
> pip install pync
> pip install websocket-client
> weechat
> ```

 I literally signed-in to say that this was the solution for #540 for Arch users.

**Comment by scottro11** [comment]
This also worked for me in June 2019.  I had tried debugging as listed in https://github.com/wee-slack/wee-slack#development but that gives me no output in buffers, and don't know much about python or weechat, so not sure what  I missed. I have found that in Fedora and FreeBSD, upgrading weechat has broken the plugin there as well, In FreeBSD, I was able to reinstall weechat-2.3 which works.  https://github.com/wee-slack/wee-slack/issues/676
